pytorch_lightning/utilities/exceptions.py

```diff
@@ -19,3 +19,8 @@ class MisconfigurationException(Exception):
 
 class DeadlockDetectedException(Exception):
     """Exception used when a deadlock has been detected and processes are being killed."""
+
+
+class ExitGracefullyException(Exception):
+    """Exception used when a ``signal.SIGUSR1` is sent to the process which triggers Lightning to create fault
+    tolerant checkpoint once reached the next reproducible part of the codebase and then exits."""
```

Comment:
how is "next reproducible part" defined? this is very vague to me

Reply:
`next reproducible part` isn't fully baked yet, but it means optimization happened (optimizer.step, schedulers), logging, etc...

Here is the current mechanism to exit from the test.

```
    if failure_on_step:
        if on_last_batch:
            if failure_on_training:
                # Breaking on first validation batch.
                # This is done to capture the random state of the validation dataloader.
                status = "EvaluationEpochLoop:advance"
            else:
                # when breaking on last batch of validation, we should exist on `run_end` val_check_interval == 1.0
                status = (
                    "TrainingEpochLoop:on_run_end" if val_check_interval == 1.0 else "TrainingEpochLoop:on_advance_end"
                )
        else:
            status = "TrainingEpochLoop:on_advance_end" if failure_on_training else "EvaluationEpochLoop:advance"
    else:
        if val_check_interval == 1.0:
            status = "TrainingEpochLoop:on_run_end"
        else:
            # `training_epoch_end` happens after `validation_epoch_end` since Lightning v1.4
            status = "TrainingEpochLoop:on_run_end" if failure_on_training else "TrainingEpochLoop:on_advance_end"
```